---
title: "Illumina_Run2_Processing"
output: html_document
editor_options: 
  chunk_output_type: console
---

#### Initial settings 
```{r}
silvadb <- 132 # set the version of the SILVA release that was used for classification
rdprel <- 16 #set the version of the RDP release that was used
fldr <- "/media/projects2/LisaM/Batch_Experiments/OTU-recovery2_AllBatch/2_Processing/" #relative location from the report project to the data.
dataname <- "Batch"
metadatafn <- "MetaData.xlsx" #relative path (from fldr) to select other Metadata files
#during the SOP workshop we'll tell people to use the pre-trimmed file in the 
#taxonomy folders on the server to have a more reproducible way of calling this.

```

```{r}
# set global chunk options: 
library(knitr)
opts_chunk$set(cache=TRUE, autodep = TRUE)
#dep_auto()

#in this way, for every chunck, if any previous chunck is changed, it will be recompiled
#given that only generally the first chunck should change, on first run the cache will be 
#rebuilt by default but subsequent downstream changes should be fine
```

#### load required packages ####
```{r}
#### load required packages ####
library(scales)   #for percent formatting
library(ggplot2)  #for adequate plotting
library(plyr)     #data wrangling (mapvalues)
suppressPackageStartupMessages(library(dplyr))    #data wrangling
library(tidyr)    #tidy data
suppressPackageStartupMessages(library(reshape2)) #for melt function
library(vegan)    #for ecological calculations
library(phyloseq) #for microbiome census data processing
library(ade4)     #for ecological calculations
library(gplots)   #for heatmap.2
library(splitstackshape) #for csplit
library(knitr)    #for simple markdown tables
library(xtable)   #for more advanced tables
library(ape)      #dealing with phylogenetic trees
library(xlsx)     #handling Excel files
library(openxlsx) #handling Excel files without Java deps
library(readxl)   #faster handling of excel files
library(data.table) #data wrangling
library(SPECIES)  #alpha diversity estimators
library(parallel) #parallel computation in R
library(purrr) #functional programming
library(DESeq2)  #normalization procedures to cope with differences in smp depth
library(NMF)
library(devtools) #nicer session info
library(viridis) # for color-blind nice color palettes
library(Phenoflow)
library(pander)

```

```{r}

#### user defined functions ####
#need to be integrated in dedicated package at https://github.ugent.be/LabMETNGS/CMETNGS_package
preformattax <- function(taxonomy)     #reformats column with taxonomy (see excelfile) and splits it into different columns
{
  #Input: mothur taxonomy with probs=T (default)
  #Output: splitted and stripped tax file for R processing (e.g. phyloseq)
  tax.good <- cSplit(taxonomy,"Taxonomy",";")
  tax.good.probs <- do.call(cbind, lapply(tax.good[,3:ncol(tax.good)], 
                                          function(x){
                                            do.call(rbind, 
                                                    strsplit(as.character(x), 
                                                             "\\((?=\\d)", 
                                                             perl = TRUE))
                                          }
  )
  )
  # tax.good.probs.nona <- subset(tax.good.probs,
  #           select=which(colSums(apply(tax.good.probs,2,is.na))==0))
  tax.final <- as.data.frame(apply(tax.good.probs,2,
                                   function(x) sub(")","",x)))
  otunames <- tax.good$OTU
  rownames(tax.final) <- otunames
  colnames(tax.final) <- c("Regnum","Prob_R","Phylum","Prob_P",
                           "Classis","Prob_C","Ordo","Prob_O",
                           "Familia","Prob_F","Genus","Prob_G")
  if(ncol(tax.final)==12){
    #for RDP & SILVA taxonomies which stop at genus level
    tax.final <- as.data.frame(mutate(tax.final,Species=NA,Prob_S=NA))
    rownames(tax.final) <- otunames
  }else{
    #for greengenes taxonomy
    colnames(tax.final)[13:14] <- c("Species","Prob_S")
    tax.final <- as.data.frame(apply(tax.final,
                                     2,
                                     function(x)sub("[a-z]__","",x)))
    tax.final$Species <- paste(tax.final$Genus,tax.final$Species) #just the species is not very sensible
  }
  #to comply to mothur 1.38 tax format
  charvectgenus <- as.character(tax.final$Genus)
  probvectgenus <- as.numeric(as.character(tax.final$Prob_G))
  nasgenuslev <- which(is.na(charvectgenus))
  for(i in nasgenuslev)
  {
    if(grepl("unclassified",as.character(tax.final$Familia[i])))
    {
      if(!is.na(as.character(tax.final$Familia[i])))
      {
        charvectgenus[i]<-as.character(tax.final$Familia[i])
      }else{
        if(grepl("unclassified",as.character(tax.final$Ordo[i])))
        {
          charvectgenus[i]<-as.character(tax.final$Ordo[i])
        }
      }
      
    }else{
      if(!is.na(as.character(tax.final$Familia[i])))
      {
        charvectgenus[i]<-paste0(as.character(tax.final$Familia[i]),"_unclassified")
      }else{
        charvectgenus[i]<-paste0(as.character(tax.final$Ordo[i]),"_unclassified")
      }
      
    }
    probvectgenus[i] <- as.numeric(as.character(tax.final$Prob_F[i]))
  }
  tax.final$Genus <- factor(charvectgenus)
  tax.final$Prob_G <- probvectgenus
  
  charvectfamilia <- as.character(tax.final$Familia)
  probvectfamilia <- as.numeric(as.character(tax.final$Prob_F))
  nasfamlev <- which(is.na(charvectfamilia))
  for(i in nasfamlev)
  {
    if(grepl("unclassified",as.character(tax.final$Ordo[i])))
    {
      if(!is.na(as.character(tax.final$Ordo[i])))
      {
        charvectfamilia[i]<-as.character(tax.final$Ordo[i])
      }else{
        if(grepl("unclassified",as.character(tax.final$Classis[i])))
        {
          charvectfamilia[i]<-as.character(tax.final$Classis[i])
        }
      }
      
    }else{
      if(!is.na(as.character(tax.final$Ordo[i])))
      {
        charvectfamilia[i]<-paste0(as.character(tax.final$Ordo[i]),"_unclassified")
      }else{
        charvectfamilia[i]<-paste0(as.character(tax.final$Classis[i]),"_unclassified")
      }
      
    }
    probvectfamilia[i] <- as.numeric(as.character(tax.final$Prob_O[i]))
  }
  tax.final$Familia <- factor(charvectfamilia)
  tax.final$Prob_F <- probvectfamilia
  return(tax.final)
}

preformattax.new <- function(taxonomy)
{
  #Input: mothur taxonomy with probs=T (default)
  #Output: splitted and stripped tax file for R processing (e.g. phyloseq)
  tax.good <- cSplit(taxonomy,"Taxonomy",";")
  tax.good.probs <- do.call(cbind, lapply(tax.good[,3:ncol(tax.good)], 
                                          function(x){
                                            do.call(rbind, 
                                                    strsplit(as.character(x), 
                                                             "\\((?=\\d)", 
                                                             perl = TRUE))
                                          }
  )
  )
  #above lies difference with regular preformattax, checks for number after bracket (line above and below)
  tax.final <- as.data.frame(apply(tax.good.probs,2,
                                   function(x) sub(")","",x)))
  
  # tax.final.noprobs <- as.data.frame(apply(tax.good.probs,2,
  #                                 function(x) sub(")","",x)))
  # tax.final <- suppressMessages(suppressWarnings(as.data.frame(sapply(tax.final.noprobs,
  #                                     function(x){mapvalues(x,from="00",to="100")}))))
  otunames <- tax.good$OTU
  #tax.final <- subset(tax.final,select=-c(OTU,Size))
  rownames(tax.final) <- otunames
  colnames(tax.final) <- c("Regnum","Prob_R","Phylum","Prob_P",
                           "Classis","Prob_C","Ordo","Prob_O",
                           "Familia","Prob_F","Genus","Prob_G")
  if(ncol(tax.final)==12){
    #for RDP & SILVA taxonomies which stop at genus level
    tax.final <- as.data.frame(mutate(tax.final,Species=NA,Prob_S=NA))
    rownames(tax.final) <- otunames
  }else{
    #for greengenes taxonomy
    colnames(tax.final)[13:14] <- c("Species","Prob_S")
    tax.final <- as.data.frame(apply(tax.final,
                                     2,
                                     function(x)sub("[a-z]__","",x)))
    tax.final$Species <- paste(tax.final$Genus,tax.final$Species) #just the species is not very sensible
  }
  #to comply to mothur 1.38 tax format
  charvectgenus <- as.character(tax.final$Genus)
  probvectgenus <- as.numeric(as.character(tax.final$Prob_G))
  nasgenuslev <- which(is.na(charvectgenus))
  for(i in nasgenuslev)
  {
    if(grepl("unclassified",as.character(tax.final$Familia[i])))
    {
      if(!is.na(as.character(tax.final$Familia[i])))
      {
        charvectgenus[i]<-as.character(tax.final$Familia[i])
      }else{
        if(grepl("unclassified",as.character(tax.final$Ordo[i])))
        {
          charvectgenus[i]<-as.character(tax.final$Ordo[i])
        }
      }
      
    }else{
      if(!is.na(as.character(tax.final$Familia[i])))
      {
        charvectgenus[i]<-paste0(as.character(tax.final$Familia[i]),"_unclassified")
      }else{
        charvectgenus[i]<-paste0(as.character(tax.final$Ordo[i]),"_unclassified")
      }
      
    }
    probvectgenus[i] <- as.numeric(as.character(tax.final$Prob_F[i]))
  }
  tax.final$Genus <- factor(charvectgenus)
  tax.final$Prob_G <- probvectgenus
  
  charvectfamilia <- as.character(tax.final$Familia)
  probvectfamilia <- as.numeric(as.character(tax.final$Prob_F))
  nasfamlev <- which(is.na(charvectfamilia))
  for(i in nasfamlev)
  {
    if(grepl("unclassified",as.character(tax.final$Ordo[i])))
    {
      if(!is.na(as.character(tax.final$Ordo[i])))
      {
        charvectfamilia[i]<-as.character(tax.final$Ordo[i])
      }else{
        if(grepl("unclassified",as.character(tax.final$Classis[i])))
        {
          charvectfamilia[i]<-as.character(tax.final$Classis[i])
        }
      }
      
    }else{
      if(!is.na(as.character(tax.final$Ordo[i])))
      {
        charvectfamilia[i]<-paste0(as.character(tax.final$Ordo[i]),"_unclassified")
      }else{
        charvectfamilia[i]<-paste0(as.character(tax.final$Classis[i]),"_unclassified")
      }
      
    }
    probvectfamilia[i] <- as.numeric(as.character(tax.final$Prob_O[i]))
  }
  tax.final$Familia <- factor(charvectfamilia)
  tax.final$Prob_F <- probvectfamilia
  return(tax.final)
}

library(CMETNGS) #will be integrated into CMETNGS package

```


#### load data 
```{r}
#### load data ####

if(.Platform$OS.type == "unix") {
  crfn <- system(paste("ls",fldr," | grep contigs.report"),intern=TRUE)
} else {
  filelist <- list.files(fldr)
  crfn <- grep(".*contigs.report",filelist,value=TRUE)
}
fn <- sub(".contigs.report","",crfn)
ini <- fread(paste(fldr,"/",crfn,sep=""),header=TRUE)
csum <- fread(paste(fldr,"/",fn,
                    ".trim.contigs.summary",sep=""),header=TRUE)
firsttrimsum <- fread(paste(fldr,"/",fn,
                            ".trim.contigs.good.summary",sep=""),
                      header=TRUE)
uniquesum <- fread(paste(fldr,"/",fn,
                         ".trim.contigs.good.unique.summary",sep=""),
                   header=TRUE)
postalnsum <- fread(paste(fldr,"/",fn,
                          ".trim.contigs.good.unique.good.filter.summary",sep=""),
                    header=TRUE)
preclussum <- fread(paste(fldr,"/",fn,
                          ".trim.contigs.good.unique.good.filter.unique.precluster.summary",sep=""),
                    header=TRUE)
postuchimeclasssum <- fread(paste(fldr,"/",fn,".trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.summary",sep=""),
                            header=TRUE)




# By default (if only one taxonomy, run the code below)
# otutaxonomy <- fread(paste(fldr,"/",fn,
# ".trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.an.unique_list.0.03.cons.taxonomy",
#             sep=""),header=TRUE)


# otherwise, we default to SILVA (for now)
otutaxonomy <- fread(paste(fldr,"/",fn,
                           ".trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.opti_mcc.0.03.cons.taxonomy",
                           sep=""),header=TRUE)

taxonomy.spl <- preformattax.new(otutaxonomy)
taxonomy.np <- taxonomy.spl %>% dplyr::select(-dplyr::contains("Prob"))    #np = no probabilities

# read in reads per sample
shared <- fread(paste(fldr,"/",fn,
                      ".trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.opti_mcc.shared",sep=""),
                header=TRUE)
shared <- as.data.frame(shared)
desgroups <- shared$Group
shared.x <- shared[,4:ncol(shared)]    # OTU-table
rownames(shared.x) <- desgroups
shared.t <- as.data.frame(t(shared.x))   # transposed: OTU's are rows, samples are columns
shared.t.ns <- shared.t[which(rowSums(shared.t)!=1),]    # ns = no singletons
sharedns <- data.frame(label=rep(0.03,ncol(shared.t.ns)),
                       Group=colnames(shared.t.ns),
                       numOtus=nrow(shared.t.ns),t(shared.t.ns))
suppressWarnings(try(write.table(x=sharedns,file=paste(fldr,"/sharedns.shared",sep=""),
                                 row.names=FALSE,quote=FALSE),silent=TRUE))


#write.xlsx2(x = shared.t, file = "Results.xlsx", sheetName = "OTU_table" )
#write.xlsx2(x = sharedns, file = "Results.xlsx",
#            sheetName = "OTU_tablenosingletons",
#            append = TRUE)

#filter taxonomy (i.e. remove singletons from taxonomy)
taxonomy.np.ns <- taxonomy.np[which(rownames(taxonomy.np) 
                                    %in% rownames(shared.t.ns)),]


```


###Read Metadata
```{r}
###Read Metadata###

metadata <- readxl::read_excel("MetaData.xlsx",sheet = "ForR")

######metadata.tibble <- readxl::read_excel(paste0(fldr,metadatafn),sheet="ForR")
######factdescs <- readxl::read_excel(paste0(fldr,metadatafn),sheet="FactDesc")

#TODO(fpkerckh) : generic naming & format for MD? 
metadata <- as.data.frame(metadata) #to avoid warnings/errors with rownames
if(is.numeric(metadata$SampleName)) #check for fully numeric sample names
{
  #metadata$SampleName <- paste(metadata$EM,metadata$EM_Conc,metadata$Factor1,metadata$Factor2,sep=" - ")
  metadata$SampleName <- as.character(metadata$SampleName)
}
rownames(metadata) <- metadata$SampleName
metadata.smpdat <- sample_data(metadata)

colnames(shared.t.ns) <- plyr::mapvalues(colnames(shared.t.ns),
                                         from=as.character(metadata$Code),
                                         to=as.character(metadata$SampleName)) #rename

#the part above maps sample names to codes that were used for sequencing at LGC, regardless of the order in the shared file

```


#### phyloseq constructors
```{r}

#### phyloseq constructors ####
#### only needed with deseq and...    not for barplots
otumat.ns <- as.matrix(shared.t.ns)
taxmat.ns <- as.matrix(taxonomy.np.ns)
OTU       <- otu_table(otumat.ns,taxa_are_rows = TRUE)
TAX       <- tax_table(taxmat.ns)
physeqobj <- phyloseq(OTU,TAX)
physeqobj.meta <- merge_phyloseq(physeqobj,metadata.smpdat)

ntaxa(physeqobj.meta)
nsamples(physeqobj.meta)
rank_names(physeqobj.meta)
sample_variables(physeqobj.meta)


TopNOTUs <- names(sort(taxa_sums(physeqobj.meta), TRUE)[1:10])
top10   <- prune_taxa(TopNOTUs, physeqobj.meta)
p <- plot_bar(top10, x="Factor1", fill="Genus")
p

```

###WNWN-filtering
```{r}

#### filtering according to WNWN ####
## prevalence = (fraction of samples in which an OTU is observed minimum 1 time)
minobs=1
prevalence <- apply(as.matrix(shared.t.ns),1,function(x,minobs){sum(x>=minobs)},minobs)/ncol(shared.t.ns) #kollom- of rijmatrix met 1 cijfer per OTU
prevalencefilter <- prevalence>0.05   #een OTU moet in minstens 5% vd stalen minstens 1x voorkomen
sharedminsingletonwnwn <- shared.t.ns[prevalencefilter,]


##Read counts per OTU should exceed 0.5 times the number of samples
sharedfilteredwnwn <- sharedminsingletonwnwn[rowSums(sharedminsingletonwnwn)>0.5*ncol(sharedminsingletonwnwn),]
# deseq normalise ==> very dependent upon design
metadata$factor1 <- factor(metadata$Factor1)
metadata$factor2 <- factor(metadata$Factor2)
deseqdata <- as.matrix(sharedfilteredwnwn +1)
# deseqdata <- DESeqDataSetFromMatrix(deseqdata,colData=metadata,design= ~ factor1 + factor2)
# deseqdata <- estimateSizeFactors(deseqdata)
# sharedfiltereddeseq <- counts(deseqdata,normalized=TRUE)
# deseqdata <- estimateDispersions(deseqdata,fitType="local")
# sharedfiltereddeseqvartransf <- varianceStabilizingTransformation(deseqdata, blind = FALSE)
# sharedfiltereddeseqvartransfmatrix <- assay(sharedfiltereddeseqvartransf)
# wnwndeseqvartransf <- getVarianceStabilizedData(deseqdata)
# jaccard_deseq <- vegdist(t(sharedfiltereddeseq),method="jaccard",binary="FALSE") # not interpretable on negative counts!

```


### Graphs controlling Mothur pipeline ###
```{r}
### Graphs controlling Mothur pipeline ###

#csum = total number of contigged reads in all samples
hist(csum$nbases,col = "lightblue",main="",xlab="Number of nucleotides")
box()

#firsttrimsum = contigs filterred for wrong length
hist(firsttrimsum$nbases,col = "lightblue",main="",xlab="Number of nucleotides")
box()

#uniquesum = filtered for unique sequences
hist(uniquesum$nbases,col = "lightblue",main="",xlab="Number of nucleotides")
box()

#postalnsum = post alignment to SILVA seed (in my case the E.coli from Jeroen Raes)
hist(postalnsum$nbases,col = "lightblue",main="",xlab="Number of nucleotides")
box()


#preclussum = after preclustering


#postuchimeclasssum = after chimera removal and removal of archaea and other non bacterial DNA
  

countdfprepr <- data.frame(uniqueseqs=c(nrow(csum),nrow(firsttrimsum),
                                        nrow(uniquesum),nrow(postalnsum),
                                        nrow(preclussum),
                                        nrow(postuchimeclasssum)),
                           totalseqs=c(nrow(csum),nrow(firsttrimsum),
                                       sum(uniquesum$numSeqs),
                                       sum(postalnsum$numSeqs),
                                       sum(postalnsum$numSeqs),
                                       sum(postuchimeclasssum$numSeqs)),
                           step=c("Contigs","Initial trim",
                                  "Unique","Alignment",
                                  "Precluster","UChime"))

countdfprepr.m <- melt(countdfprepr)
p <- ggplot(aes(x=step,y=value,fill=variable),
            data=countdfprepr.m) +
      geom_bar(stat="identity",position="dodge") +
      xlab("Nucleotide") + 
      ylab("Number of nucleotides") + theme_bw() +
      scale_fill_discrete(labels=c("Unique",
                                   "Total"),
                          name="") + 
      scale_x_discrete(limits=c("Contigs",
                                "Initial trim",
                                "Unique",
                                "Alignment",
                                "Precluster",
                                "UChime")) +
      theme(legend.title=element_blank())
p
write.xlsx2(x = countdfprepr.m, file = "Results.xlsx", 
            sheetName = "filteredReadcounts", append = TRUE)

```


```{r}
# groupcounts = dataframe with all the sample names
# this is to check if in every sample the amount of reads is sufficient & approximately the same

groupcounts <- data.frame(Reads = sort(colSums(shared.t.ns),
                                       decreasing=TRUE))
kable(groupcounts,
      caption = "Number of reads for all groups after removing singletons")



write.xlsx2(x = groupcounts, file = "Results.xlsx",
            sheetName = "groupcounts", append = TRUE)

```


```{r}
groupcountdf <- data.frame(groupcounts,samplename=rownames(groupcounts))
metadatadf <- data.frame(metadata,samplename=rownames(metadata))
countsdfplot <- dplyr::full_join(groupcountdf,metadatadf,by="samplename")
countsdfplot$Factor1 <- as.factor(countsdfplot$Factor1)
countsdfplot$Factor2 <- as.factor(countsdfplot$Factor2)
gglibsize <- ggplot(countsdfplot,aes(x=Factor1,y=Reads,fill=Factor1))
gglibsize + geom_boxplot() + facet_grid(~Factor2,drop=TRUE) + 
  theme_bw() +
  theme(axis.title.x = element_blank(),axis.text.x = element_blank()) +
  scale_fill_discrete(name=factdescs$Desc[1])+
  ggtitle(paste("Library size by",factdescs$Desc[1],"and",factdescs$Desc[2]))
  

```

# Own attempt at bargraphs


```{r}
library(vegan)
#sorteer shared.t.ns volgens samplenummer
shared.t.ns <- shared.t.ns[, as.character(metadata$SampleName)]

# calculation of relative abundances (instead of absolute reads)
shared.t.ns.relabun <- decostand(data.frame(shared.t.ns[1:255]), 2, method = "total")
colnames(shared.t.ns.relabun) <- colnames(shared.t.ns)
```


### GENUS
```{r}
# add taxonomy column 
shared.t.ns.relabun$Genus <- taxonomy.np.ns$Genus

#summate relative abundances in each sample per genus
shared.t.ns.relabun.Gr <- aggregate(shared.t.ns.relabun[1:255], by=list(Category=shared.t.ns.relabun$Genus), FUN=sum)

#filter out 8 most abundant genusses
shared.t.ns.relabun.Gr$Sums <- rowSums(shared.t.ns.relabun.Gr[2:256])
shared.t.ns.relabun.Gr <- shared.t.ns.relabun.Gr[order(-shared.t.ns.relabun.Gr$Sums),]
shared.t.ns.relabun.Gr.top8 <- shared.t.ns.relabun.Gr[1:8,]
shared.t.ns.relabun.Gr.top8[9,2:257] <- colSums(shared.t.ns.relabun.Gr[9:281,2:257])

#adding "Other" on nineth place of first column
shared.t.ns.relabun.Gr.top8$Category <- as.character(shared.t.ns.relabun.Gr.top8$Category)
shared.t.ns.relabun.Gr.top8$Category[9] <- c("Other")
shared.t.ns.relabun.Gr.top8$Category <- as.factor(shared.t.ns.relabun.Gr.top8$Category)

shared.t.ns.relabun.Gr.top8$Sums <- NULL

#rearrange data-table in order to be able to add metadata
shared.t.ns.relabun.Gr.top8.R <- tidyr::gather(shared.t.ns.relabun.Gr.top8,key = "SampleName", value = "RelativeAbundance",2:256)

#add metadata
shared.merge.gen <- merge(shared.t.ns.relabun.Gr.top8.R,metadata[2:6], by = "SampleName")
shared.merge.gen <- dplyr::arrange(shared.merge.gen,as.numeric(shared.merge.gen$SampleName))
shared.merge.gen <- unite(shared.merge.gen, Treatment, c(EM, EM_Conc), remove=FALSE, sep =' - ')

#plot bargraphs

BarOrder <- c("Other","Prevotella","Parabacteroides","Lachnospiraceae_unclassified","Faecalibacterium","Enterobacteriaceae_unclassified","Blautia","Bacteroides","Alistipes" )

X.order <- c("CMC - 0.005","CMC - 0.05","CMC - 0.5","P80 - 0.005", "P80 - 0.05","P80 - 0.5","SoyLec - 0.005", "SoyLec - 0.05","SoyLec - 0.5","SL - 0.005","SL - 0.05","SL - 0.5","RL - 0.005","RL - 0.05","RL - 0.5", "Control - 0")

X.order.RL <- c("Control - 0","RL - 0.005","RL - 0.05","RL - 0.5")

#position_fill(reverse = TRUE)
barplot.genus.me <-ggplot(data=shared.merge.gen, aes(x= Treatment, y=RelativeAbundance, fill = factor(Category, levels=as.character(shared.t.ns.relabun.Gr.top8$Category)))) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE))+  
  facet_grid(Factor2~Factor1)+
  theme_minimal()+ 
  #theme(axis.text.x = element_text(size=10), legend.title=element_text(size=10), legend.text=element_text(size=7))
  scale_fill_brewer(palette="Set1")+
  ggtitle("Batch Experiments - Barplot Genus")+
  ylab("Relative Abundance")+
  theme(axis.text.x=element_text(size =8, angle=90, hjust=1))+
  labs(fill = "Top 8 Genus")+
  scale_x_discrete(limits=X.order)
barplot.genus.me

barplot.genus.me.RL <-ggplot(data=shared.merge.gen, aes(x= Treatment, y=RelativeAbundance, fill = factor(Category, levels=as.character(shared.t.ns.relabun.Gr.top8$Category)))) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE))+  
  facet_grid(Factor2~Factor1)+
  theme_minimal(base_size = 10)+ 
  scale_fill_brewer(palette="Set1")+
  ggtitle("Batch Experiments - Barplot Genus")+
  ylab("Relative Abundance")+
  theme(axis.text.x=element_text(angle=90, hjust=1))+
  labs(fill = "Top 8 Genus")+
  scale_x_discrete(limits=c("Control - 0","RL - 0.005","RL - 0.05","RL - 0.5"))
barplot.genus.me.RL

#levels (as.factor(shared.merge.gen$Treatment))


#plot only Entero's
shared.merge.gen.Ent <- shared.merge.gen %>%  filter(as.factor(shared.merge.gen$Category) == "Enterobacteriaceae_unclassified")


Abs.Ent <- shared.t.ns[2,]
Rel.Ent <- shared.t.ns.relabun[2,1:255]

Ent.data <- rbind(Abs.Ent,Rel.Ent)
rownames(Ent.data) <- c("Abs.Enterobacteriaceae","Rel.Enterobacteriaceae")
Ent.data <- data.frame(t(Ent.data))
Ent.data$SampleName <- rownames(Ent.data)

#add metadata
Ent.data.merge <- merge(Ent.data,metadata[2:6], by = "SampleName")
Ent.data.merge <- dplyr::arrange(Ent.data.merge,as.numeric(Ent.data.merge$SampleName))
Ent.data.merge <- unite(Ent.data.merge, Treatment, c(EM, EM_Conc), remove=FALSE, sep =' - ')
Ent.data.merge$Rel.Enterobacteriaceae <-Ent.data.merge$Rel.Enterobacteriaceae *3*10000

#rearrange data-table in order to be able to add metadata
Ent.data.merge.R <- tidyr::gather(Ent.data.merge,key = "RelAbs", value =Abundance,c("Abs.Enterobacteriaceae","Rel.Enterobacteriaceae"))

barplot.Entero <-ggplot(data=Ent.data.merge.R, aes(x= Treatment,y = Abundance, fill = as.factor(RelAbs))) +
  geom_bar(stat ="identity", position ="dodge")+  
  scale_y_continuous(sec.axis = sec_axis(trans = ~./(3*10000),name = "Rel.Abundance"))+
  facet_grid(Factor2~Factor1)+
  theme_minimal()+ 
  scale_fill_brewer(palette="Set1")+
  ggtitle("Batch Experiments - Barplot Enterobacteriaceae")+
  ylab("Relative Abundance")+
  theme(axis.text.x=element_text(angle=90, hjust=1))+
  labs(fill = "Abundance")+
  scale_x_discrete(limits=X.order)
barplot.Entero

```

### FAMILY 
```{r}
### FAMILY ###
# add taxonomy column 
shared.t.ns.relabun$Familia <- taxonomy.np.ns$Familia

#summate relative abundances in each sample per genus
shared.t.ns.relabun.Gr.fam <- aggregate(shared.t.ns.relabun[1:255], by=list(Category=shared.t.ns.relabun$Familia), FUN=sum)

#filter out 8 most abundant genusses
shared.t.ns.relabun.Gr.fam$Sums <- rowSums (shared.t.ns.relabun.Gr.fam[2:256])
shared.t.ns.relabun.Gr.fam <- shared.t.ns.relabun.Gr.fam[order(-shared.t.ns.relabun.Gr.fam$Sums),]
shared.t.ns.relabun.Gr.fam.top8 <- shared.t.ns.relabun.Gr.fam[1:8,]
Nfam = nrow(shared.t.ns.relabun.Gr.fam)
shared.t.ns.relabun.Gr.fam.top8[9,2:257] <- colSums(shared.t.ns.relabun.Gr.fam[9:Nfam,2:257])  

#adding "Other" on nineth place of first column
shared.t.ns.relabun.Gr.fam.top8$Category <- as.character(shared.t.ns.relabun.Gr.fam.top8$Category)
shared.t.ns.relabun.Gr.fam.top8$Category[9] <- c("Other")
shared.t.ns.relabun.Gr.fam.top8$Category <- as.factor(shared.t.ns.relabun.Gr.fam.top8$Category)

#remove Sums-column (was only used for picking out most abundant)
shared.t.ns.relabun.Gr.fam.top8$Sums <- NULL

#rearrange data-table in order to be able to add metadata
shared.t.ns.relabun.Gr.fam.top8.R <- tidyr::gather(shared.t.ns.relabun.Gr.fam.top8,key = "SampleName", value = "RelativeAbundance",2:256)

#add metadata
shared.merge.Fam <- merge(shared.t.ns.relabun.Gr.fam.top8.R,metadata[2:6], by = "SampleName")
shared.merge.Fam <- dplyr::arrange(shared.merge.Fam,as.numeric(shared.merge.Fam$SampleName))
shared.merge.Fam <- unite(shared.merge.Fam, Treatment, c(EM, EM_Conc), remove=FALSE, sep =' - ')

#plot bargraphs

X.order <- c("CMC - 0.005","CMC - 0.05","CMC - 0.5","P80 - 0.005", "P80 - 0.05","P80 - 0.5","SoyLec - 0.005", "SoyLec - 0.05","SoyLec - 0.5","SL - 0.005","SL - 0.05","SL - 0.5","RL - 0.005","RL - 0.05","RL - 0.5", "Control - 0")

X.order.RL <- c("Control - 0","RL - 0.005","RL - 0.05","RL - 0.5")

barplot.fam.me <-ggplot(data=shared.merge.Fam, aes(x= Treatment, y=RelativeAbundance, fill = factor(Category, levels=as.character(shared.t.ns.relabun.Gr.fam.top8$Category)))) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE))+  
  facet_grid(Factor2~Factor1)+
  theme_minimal(base_size = 10)+ 
  scale_fill_brewer(palette="Set1")+
  ggtitle("Batch Experiments - Barplot Family")+
  ylab("Relative Abundance")+
  theme(axis.text.x=element_text(angle=90, hjust=1))+
  labs(fill = "Top 8 Families")+
  scale_x_discrete(limits=X.order)
barplot.fam.me

barplot.fam.me.RL <-ggplot(data=shared.merge.Fam, aes(x= Treatment, y=RelativeAbundance, fill = factor(Category, levels=as.character(shared.t.ns.relabun.Gr.fam.top8$Category)))) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE))+  
  facet_grid(Factor2~Factor1)+
  theme_minimal(base_size = 10)+ 
  scale_fill_brewer(palette="Set1")+
  ggtitle("Batch Experiments - Barplot Family")+
  ylab("Relative Abundance")+
  theme(axis.text.x=element_text(angle=90, hjust=1))+
  labs(fill = "Top 8 Families")+
  scale_x_discrete(limits=X.order.RL)
barplot.fam.me.RL





```

```{r}
### ORDO


### CLASSIS
```

### PHYLUM
```{r}
# add taxonomy column 
shared.t.ns.relabun$Phylum <- taxonomy.np.ns$Phylum

#summate relative abundances in each sample per genus
shared.t.ns.relabun.Gr.phyl <- aggregate(shared.t.ns.relabun[1:255], by=list(Category=shared.t.ns.relabun$Phylum), FUN=sum)

#filter out 8 most abundant genusses
shared.t.ns.relabun.Gr.phyl$Sums <- rowSums(shared.t.ns.relabun.Gr.phyl[2:256])
shared.t.ns.relabun.Gr.phyl <- shared.t.ns.relabun.Gr.phyl[order(-shared.t.ns.relabun.Gr.phyl$Sums),]
shared.t.ns.relabun.Gr.phyl.top8 <- shared.t.ns.relabun.Gr.phyl[1:8,]
Nphyl = nrow(shared.t.ns.relabun.Gr.phyl)
shared.t.ns.relabun.Gr.phyl.top8[9,2:257] <- colSums(shared.t.ns.relabun.Gr.phyl[9:Nphyl,2:257])

#adding "Other" on nineth place of first column
shared.t.ns.relabun.Gr.phyl.top8$Category <- as.character(shared.t.ns.relabun.Gr.phyl.top8$Category)
shared.t.ns.relabun.Gr.phyl.top8$Category[9] <- c("Other")
shared.t.ns.relabun.Gr.phyl.top8$Category <- as.factor(shared.t.ns.relabun.Gr.phyl.top8$Category)

shared.t.ns.relabun.Gr.phyl.top8$Sums <- NULL

#rearrange data-table in order to be able to add metadata
shared.t.ns.relabun.Gr.phyl.top8.R <- tidyr::gather(shared.t.ns.relabun.Gr.phyl.top8,key = "SampleName", value = "RelativeAbundance",2:256)

#add metadata
shared.merge.phyl <- merge(shared.t.ns.relabun.Gr.phyl.top8.R,metadata[2:6], by = "SampleName")
shared.merge.phyl <- dplyr::arrange(shared.merge.phyl,as.numeric(shared.merge.phyl$SampleName))
shared.merge.phyl <- unite(shared.merge.phyl, Treatment, c(EM, EM_Conc), remove=FALSE, sep =' - ')

#plot bargraphs

X.order <- c("CMC - 0.005","CMC - 0.05","CMC - 0.5","P80 - 0.005", "P80 - 0.05","P80 - 0.5","SoyLec - 0.005", "SoyLec - 0.05","SoyLec - 0.5","SL - 0.005","SL - 0.05","SL - 0.5","RL - 0.005","RL - 0.05","RL - 0.5", "Control - 0")

X.order.RL <- c("Control - 0","RL - 0.005","RL - 0.05","RL - 0.5")

#position_fill(reverse = TRUE)

barplot.phylum.me <-ggplot(data=shared.merge.phyl, aes(x= Treatment, y=RelativeAbundance, fill = factor(Category, levels=as.character(shared.t.ns.relabun.Gr.phyl.top8$Category)))) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE))+  
  facet_grid(Factor2~Factor1)+
  theme_minimal(base_size = 10)+ 
  scale_fill_brewer(palette="Set1")+
  ggtitle("Batch Experiments - Barplot Phylum")+
  ylab("Relative Abundance")+
  theme(axis.text.x=element_text(angle=90, hjust=1))+
  labs(fill = "Top 8 Phyla")+
  scale_x_discrete(limits=X.order)
barplot.phylum.me

barplot.phylum.me.RL <-ggplot(data=shared.merge.phyl, aes(x= Treatment, y=RelativeAbundance, fill = factor(Category, levels=as.character(shared.t.ns.relabun.Gr.phyl.top8$Category)))) +
  geom_bar(stat="identity", position = position_fill(reverse = TRUE))+  
  facet_grid(Factor2~Factor1)+
  theme_minimal(base_size = 10)+ 
  scale_fill_brewer(palette="Set1")+
  ggtitle("Batch Experiments - Barplot Phylum")+
  ylab("Relative Abundance")+
  theme(axis.text.x=element_text(angle=90, hjust=1))+
  labs(fill = "Top 8 Phyla")+
  scale_x_discrete(limits=c("Control - 0","RL - 0.005","RL - 0.05","RL - 0.5"))
barplot.phylum.me.RL


### write out plots into pdf file
pdf("Plot_Batch_Illumina_Run2_Classification.pdf",width = 12, height = 4, pointsize = 10)
    barplot.genus.me
    barplot.fam.me
    barplot.phylum.me
    barplot.genus.me.RL
    barplot.fam.me.RL
    barplot.phylum.me.RL
dev.off()
    
png ("Plot_Batch_Illumina_Run2_Classif_Genus.png", height = 1200, width = 2200, res = 150)
   barplot.genus.me
   dev.off()
png ("Plot_Batch_Illumina_Run2_Classif_Family.png",height = 1200, width = 2200, res = 150)
    barplot.fam.me
png ("Plot_Batch_Illumina_Run2_Classif_Phylum.png", height = 1200, width = 2200, res = 150)
    barplot.phylum.me
    dev.off()
png ("Plot_Batch_Illumina_Run2_Classif_Gen_RL.png",height = 1200, width = 2200, res = 150)
    barplot.genus.me.RL
    dev.off()
png ("Plot_Batch_Illumina_Run2_Classif_Fam_RL.png",height = 1200, width = 2200, res = 150)
    barplot.fam.me.RL
    dev.off()
png ("Plot_Batch_Illumina_Run2_Classif_Phyl_RL.png",height = 1200, width = 2200, res = 150)
    barplot.phylum.me.RL
dev.off()


```


### Classification Plots
```{r}
#TODO(fpkerckh) : make it possible to increase n 
shared.t.ns <- shared.t.ns[, as.character(metadata$SampleName)]
shared.ns <- data.frame(t(shared.t.ns))
shared.ns[,256:257] <-c(metadata$Factor1,metadata$Factor2)
metadata.barplot <- metadata
metadata.barplot$Factor1 <- c("Donor")
metadata.barplot$Factor2 <- c("Timepoint")


barplot.genus <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                        topn=8,taxlevel="Genus",shared.abs=TRUE,
                                        tax.prob=FALSE,samples=dataname,
                                        plot=TRUE)
            + theme_minimal() + facet_grid(Factor2~Factor1)

barplot.genus$ggplotdf <- merge(barplot.genus$ggplotdf,metadata[2:6],by= SampleName)

barplot.genus$ggplotobj[,4:5] <-strsplit(barplot.genus$ggplotobj[,4],



                                         
                                         
                                         
                                         
barplot.genus$ggplotdf
aes(x = paste(metadata.smpdat$EM,metadata.smpdat$EM_Conc,sep=" "))

p <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns) + facet_grid(~Factor1, drop = TRUE)
   
 
 
cat('\n\n') #from https://groups.google.com/forum/#!topic/knitr/MJVLiVyGCro
barplot.genusviridis <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                        topn=8,taxlevel="Genus",shared.abs=TRUE,
                                        tax.prob=FALSE,samples=dataname, viridfill = TRUE,
                                        plot=TRUE)

highqualgraphR(barplot.genus$ggplotobj,filename = "Barplot_genus",
               extension = c("png","tiff"),pointsize=12)
highqualgraphR(barplot.genusviridis$ggplotobj,filename = "Barplot_genus_viridis",
               extension = c("png","tiff"),pointsize=12)
genustopexp <- barplot.genus$ggplotdf  %>% dplyr::select(-Var2o)  %>% tidyr::spread(Var2,value)

write.xlsx2(x = genustopexp, file = "Results.xlsx",
            sheetName = "genusTop", append = TRUE)
#here you can order the samples in an order which you consider appropriate, using +scale_x_discrete: e.g.:
#barplot.genus$ggplotobj + scale_x_discrete(limits=c("T_Inoculum","M_OLR5_Start","T_OLR5_Start",
                                                          #"M_OLR5_End","T_OLR5_End","M_OLR10_End","T_OLR10_End",
#                                                          "M_OLR10_Biofilm","T_OLR10_Biofilm")) 
cat('\n\n')
barplot.family <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                               topn=8,taxlevel="Familia",shared.abs=TRUE,
                                               tax.prob=FALSE,samples=dataname,plot=TRUE)



cat('\n\n')
barplot.famviridis <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                               topn=8,taxlevel="Familia",shared.abs=TRUE,
                                               tax.prob=FALSE,samples=dataname,viridfill = TRUE,
                                             plot=TRUE)


familytopexp <- barplot.family$ggplotdf  %>% dplyr::select(-Var2o)  %>% tidyr::spread(Var2,value)
write.xlsx2(x = familytopexp, file = "Results.xlsx",
            sheetName = "familyTop", append = TRUE)
highqualgraphR(barplot.family$ggplotobj,filename = "Barplot_family",
               extension = c("png","tiff"),pointsize=12)
highqualgraphR(barplot.famviridis$ggplotobj,filename = "Barplot_family_viridis",
               extension = c("png","tiff"),pointsize=12)
cat('\n\n')
barplot.ordo <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                           topn=8,taxlevel="Ordo",shared.abs=TRUE,
                                           tax.prob=FALSE,samples=dataname,plot=TRUE)
cat('\n\n')
barplot.ordviridis <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                           topn=8,taxlevel="Ordo",shared.abs=TRUE,
                                           tax.prob=FALSE,samples=dataname, viridfill=TRUE,
                                           plot=TRUE)

highqualgraphR(barplot.ordo$ggplotobj,filename = "Barplot_ordo",
               extension = c("png","tiff"))
highqualgraphR(barplot.ordviridis$ggplotobj,filename = "Barplot_ordo_viridis",
               extension = c("png","tiff"),pointsize=12)
ordotopexp <- barplot.ordo$ggplotdf  %>% dplyr::select(-Var2o)  %>% tidyr::spread(Var2,value)
write.xlsx2(x = ordotopexp, file = "Results.xlsx",
            sheetName = "ordoTop", append = TRUE)

cat('\n\n')
barplot.phylum <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                           topn=8,taxlevel="Phylum",shared.abs=TRUE,
                                           tax.prob=FALSE,samples=dataname,plot=TRUE)
cat('\n\n')
barplot.phylumviridis <- makebargraphrawggplot2(tax=taxonomy.np.ns,shared=shared.t.ns,
                                           topn=8,taxlevel="Phylum",shared.abs=TRUE,
                                           tax.prob=FALSE,samples=dataname, viridfill = TRUE,
                                           plot=TRUE)
highqualgraphR(barplot.phylum$ggplotobj,filename = "Barplot_phylum",
               extension = c("png","tiff"))
highqualgraphR(barplot.phylumviridis$ggplotobj,filename = "Barplot_phylum_viridis",
               extension = c("png","tiff"),pointsize=12)
phylumtopexp <- barplot.phylum$ggplotdf  %>% dplyr::select(-Var2o)  %>% tidyr::spread(Var2,value)
write.xlsx2(x = phylumtopexp, file = "Results.xlsx",
            sheetName = "phylumTop", append = TRUE)

```

##stacked bargraphs
```{r}
TopNOTUs <- names(sort(taxa_sums(physeqobj), TRUE)[1:12])
physeqobj12  <- prune_taxa(TopNOTUs, physeqobj)
relabsphyseq <- transform_sample_counts(physeqobj, function(x) x/sum(x))
TopNOTUs.relab <- names(sort(taxa_sums(relabsphyseq), TRUE)[1:12])
physeqobj12.relab  <- prune_taxa(TopNOTUs.relab, relabsphyseq)

phygen <- plot_bar(physeqobj12, fill="Genus")
phygen + labs(title=paste("Taxonomic distribution ",dataname," samples",sep="")) +
         scale_x_discrete(limits=c(metadata$SampleName))
phygenp <- phygen + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""))+
         scale_x_discrete(limits=c(metadata$SampleName))

highqualgraphR(phygenp,filename = "Barplot_genus_phyloseq_reads",
               extension = c("png","tiff"),pointsize = 12)
cat('\n\n')
phygen + scale_fill_viridis(discrete = TRUE)
phygenvir <- phygen + scale_fill_viridis(discrete = TRUE)
highqualgraphR(phygenvir,filename = "Barplot_genus_phyloseq_reads_viridis",
               extension = c("png","tiff"),pointsize = 12)

cat('\n\n')
phyfam <- plot_bar(physeqobj12, fill="Familia")
phyfam + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""))+
         scale_x_discrete(limits=c(metadata$SampleName))
phyfamp <- phyfam + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""))+
         scale_x_discrete(limits=c(metadata$SampleName))
highqualgraphR(phyfamp,filename = "Barplot_family_phyloseq_absolute",extension = c("png"))

phyord <- plot_bar(physeqobj12, fill="Ordo")
phyord + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""))+
         scale_x_discrete(limits=c(metadata$SampleName))
phyordp <- phyord + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""))+
         scale_x_discrete(limits=c(metadata$SampleName))
highqualgraphR(phyordp,filename = "Barplot_ordo_phyloseq_absolute",extension = c("png"))

phygen.rel <- plot_bar(physeqobj12.relab, fill="Genus")
phygen.rel + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""),
                  y="Relative abundance")+
         scale_x_discrete(limits=c(metadata$SampleName))
phygenp.rel <- phygen.rel + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""),
                                 y="Relative abundance")+
         scale_x_discrete(limits=c(metadata$SampleName))
highqualgraphR(phygenp.rel,filename = "Barplot_genus_phyloseq_relative",extension = c("png"))


phyfam.rel <- plot_bar(physeqobj12.relab, fill="Familia")
phyfam.rel + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""),
                  y="Relative abundance")+
         scale_x_discrete(limits=c(metadata$SampleName))
phyfamp.rel <- phyfam.rel + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""),
                  y="Relative abundance")+
         scale_x_discrete(limits=c(metadata$SampleName))
highqualgraphR(phyfamp.rel,filename = "Barplot_family_phyloseq_relative",extension = c("png"))

phyord.rel <- plot_bar(physeqobj12.relab, fill="Ordo")
phyord.rel + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""),
                  y="Relative abundance")+
         scale_x_discrete(limits=c(metadata$SampleName))
phyordp.rel <- phyord.rel + labs(title=paste("Taxonomic distribution ",dataname," samples",sep=""),
                  y="Relative abundance")+
         scale_x_discrete(limits=c(metadata$SampleName))
highqualgraphR(phyordp.rel,filename = "Barplot_ordo_phyloseq_relative",extension = c("png"))

```

##clustered heatmap
```{r}
relabsphyseq <- transform_sample_counts(physeqobj, function(x) x/sum(x))
Top24OTUs.relab <- names(sort(taxa_sums(relabsphyseq), TRUE)[1:24])
physeqobj24.relab  <- prune_taxa(Top24OTUs.relab, relabsphyseq)
relab24mat <- as.matrix(otu_table(physeqobj24.relab))
rownames(relab24mat) <- as.character(as.data.frame(tax_table(physeqobj24.relab))$Genus)
my_palette <- colorRampPalette(c("black","green" ,"red"))(n = 409)
# this colors can be selected from a hexadecimal color picker
# e.g. https://www.google.be/search?hl=en&q=hex+color+picker&gws_rd=ssl

#heatmap.2(relab24mat,col=brewer.pal(9,name="Reds"))

col_breaks = c(seq(0,0.050,length=200),seq(0.051,0.10,length=100),
               seq(0.11,0.15,length=50),seq(0.151,0.20,length=50),
               seq(0.21,max(relab24mat),length=10))

heatmap.2(relab24mat, main = "Genus level heatmap",
          notecol="black", density.info="none",
          trace="none", margins =c(10,10), col=my_palette, 
          dendrogram="column", breaks=col_breaks,
          lmat=rbind(c(3,4),c(1,2)),lwi = c(3,1),
          Rowv = FALSE, Colv = TRUE)

```


```{r}
physeqobj.meta.subs <- prune_taxa(names(sort(taxa_sums(physeqobj.meta),
                                             TRUE)[1:100]),
                                  physeqobj.meta)
plot_heatmap(physeqobj.meta.subs, sample.label="SampleName",
             taxa.label="Genus",method = "NMDS",distance = "jaccard",
             low="#FFFFCC", high="#000033", na.value="white",trans = log_trans(10))

```


```{r}
topn <- 20
wnwnsh <- sharedfilteredwnwn
wnwnsh.prop <- decostand(wnwnsh,method = "total",MARGIN=2)
shns <- shared.t.ns
shns.prop <- decostand(shns,method = "total",MARGIN=2)
shns.log <- decostand(shns,method = "log",logbase=10)
shns.cs <- shns.prop*as.numeric(quantile(colSums(shared.t.ns),.75)) #other quantile then max see: #boxplot(colSums(shared.t.ns))
wnwnsh.log <- decostand(wnwnsh,method = "log",logbase=10)

metadata.colsidecols <- metadata %>% dplyr::select(c(Factor1,Factor2)) %>%
                                      transmute(MAP=Factor1,
                                                Day=as.factor(Factor2))
rownames(metadata.colsidecols) <- metadata$SampleName
md.colsidecols.ordered <- metadata.colsidecols[colnames(wnwnsh),]

wnwnsh.top <- wnwnsh %>% dplyr::mutate(otu=rownames(wnwnsh)) %>% 
                                     top_n(topn,rowSums(wnwnsh[1:(ncol(wnwnsh)-1)]))
rownames(wnwnsh.top) <- wnwnsh.top$otu
wnwnsh.top <- wnwnsh.top %>% dplyr::select(-c(otu))
wnwnsh.top.clust <- hclust(vegdist(t(wnwnsh.top),"jaccard",binary = FALSE),method="ward.D2")

NMF::aheatmap(wnwnsh.top,border_color="white",Rowv=NA,Colv=wnwnsh.top.clust$order,
              annCol=md.colsidecols.ordered, main="WNWN-filtered abundance-based jaccard Ward D2")

wnwnsh.prop.top <- wnwnsh.prop %>% dplyr::mutate(otu=rownames(wnwnsh.prop)) %>% 
                                     top_n(topn,rowSums(wnwnsh.prop[1:(ncol(wnwnsh.prop)-1)]))
rownames(wnwnsh.prop.top) <- wnwnsh.prop.top$otu
wnwnsh.prop.top <- wnwnsh.prop.top %>% dplyr::select(-c(otu))
wnwnsh.prop.top.clust <- hclust(vegdist(t(wnwnsh.prop.top),"jaccard",binary = FALSE),method="ward.D2")

NMF::aheatmap(wnwnsh.prop.top,border_color="white",Rowv=NA,Colv=wnwnsh.prop.top.clust$order,
              annCol=md.colsidecols.ordered,
              main="WNWN-filtered abundance-based jaccard Ward D2 relative abundances")


```


```{r}


```


```{r}


```


```{r}


```


```{r}


```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
